packages/gatsby/src/utils/api-runner-node.js

```diff
@@ -221,174 +221,164 @@ const runAPI = (plugin, api, args, activity) => {
 
-    apiSpan.setTag(`api`, api)
-    _.forEach(args.traceTags, (value, key) => {
-      apiSpan.setTag(key, value)
-    })
+  const { parentSpan, traceId, traceTags, waitForCascadingActions } = args
 
-    const plugins = store.getState().flattenedPlugins
+  const apiSpanArgs = parentSpan ? { childOf: parentSpan } : {}
+  const apiSpan = tracer.startSpan(`run-api`, apiSpanArgs)
 
-    // Get the list of plugins that implement this API.
-    // Also: Break infinite loops. Sometimes a plugin will implement an API and
-    // call an action which will trigger the same API being called.
-    // `onCreatePage` is the only example right now. In these cases, we should
-    // avoid calling the originating plugin again.
-    const implementingPlugins = plugins.filter(
-      plugin => plugin.nodeAPIs.includes(api) && plugin.name !== pluginSource
-    )
+  apiSpan.setTag(`api`, api)
+  _.forEach(traceTags, (value, key) => {
+    apiSpan.setTag(key, value)
+  })
 
-    const apiRunInstance = {
-      api,
-      args,
-      pluginSource,
-      resolve,
-      span: apiSpan,
-      startTime: new Date().toJSON(),
-      traceId: args.traceId,
-    }
+  const plugins = store.getState().flattenedPlugins
+
+  // Get the list of plugins that implement this API.
+  // Also: Break infinite loops. Sometimes a plugin will implement an API and
+  // call an action which will trigger the same API being called.
+  // `onCreatePage` is the only example right now. In these cases, we should
+  // avoid calling the originating plugin again.
+  const implementingPlugins = plugins.filter(
+    plugin => plugin.nodeAPIs.includes(api) && plugin.name !== pluginSource
+  )
+
+  const apiRunInstance = {
+    api,
+    resolve,
+    span: apiSpan,
+    traceId,
+  }
 
-    // Generate IDs for api runs. Most IDs we generate from the args
-    // but some API calls can have very large argument objects so we
-    // have special ways of generating IDs for those to avoid stringifying
-    // large objects.
-    let id
-    if (api === `setFieldsOnGraphQLNodeType`) {
-      id = `${api}${apiRunInstance.startTime}${args.type.name}${args.traceId}`
-    } else if (api === `onCreateNode`) {
-      id = `${api}${apiRunInstance.startTime}${args.node.internal.contentDigest}${args.traceId}`
-    } else if (api === `preprocessSource`) {
-      id = `${api}${apiRunInstance.startTime}${args.filename}${args.traceId}`
-    } else if (api === `onCreatePage`) {
-      id = `${api}${apiRunInstance.startTime}${args.page.path}${args.traceId}`
-    } else {
-      // When tracing is turned on, the `args` object will have a
-      // `parentSpan` field that can be quite large. So we omit it
-      // before calling stringify
-      const argsJson = JSON.stringify(_.omit(args, `parentSpan`))
-      id = `${api}|${apiRunInstance.startTime}|${apiRunInstance.traceId}|${argsJson}`
-    }
-    apiRunInstance.id = id
+  if (waitForCascadingActions) {
```

Comment:
just so I understand. `waitForCascadingActions` is used to make sure all api calls of a certain id are done. For example createPages?

Reply:
Yes I believe so. Basically a plugin can recursively cause new jobs to be ran and we would want to wait for all those jobs to finish before marking the root plugin call as resolved. (I did remove that mechanic once but it's just not the same so I left that semantic as is)